calibrators/docs/tutorials/tag_based_pnp_calibrator.md

```diff
@@ -0,0 +1,157 @@
+
+The output in the console should be as follows (variations in the numerical results may occur due to the nature of ROS):
+
+```text
+[tag_based_pnp_calibrator_node]:  Initial reprojection error=15.41
+[tag_based_pnp_calibrator_node]:  Current reprojection error=1.09
+[tag_based_pnp_calibrator_node]:  Filtered reprojection error=1.10
+```
+
+## Results
+
+After the calibration process finishes, the `sensor_calibration_manager` will display the results in the UI and allow the user to save the calibration data to a file.
+
+In the UI of the X2 project, three different TF trees are displayed: `Initial TF Tree`, `Calibration Tree`, and `Final TF Tree`.
+
+- The `Initial TF Tree` presents the initial TF connections between sensors needed for calibration.
+- The `Calibration Tree` shows the calibrated transformation between sensors, in this tutorial, `camera6/camera_optical_link` to `pandar_40p_front`.
+- The `Final TF Tree` depicts the TF tree after incorporating the updated calibrated transformation. As autoware utilizes the concept of [sensor_kit](https://autowarefoundation.github.io/autoware-documentation/main/how-to-guides/integrating-autoware/creating-vehicle-and-sensor-model/creating-sensor-model/), the final transformation required to comply to the specifications is `front_unit_base_link` to `camera6/camera_link`. This transformation is performed by the [calibrator interface](../../sensor_calibration_manager/sensor_calibration_manager/calibrators/x2/tag_based_pnp_calibrator.py) related to this project. The red arrow indicates that the final transformation changed after the calibration process.
+
+<p align="center">
+    <img src="../images/tag_based_pnp_calibrator/menu4.jpg" alt="menu4" width="500">
+</p>
+
+During or at the end of the calibration, the user can modify the `visualization options` on the right side of the `image view`. Setting the `Marker size (m)` to `0.04` and setting the `PC subsample factor` to `1` can help the user visualize the projected pointcloud on the image.
+
+<p align="center">
+    <img src="../images/tag_based_pnp_calibrator/visualization_bar.jpg"  alt="visualization_bar" width="200">
+</p>
+
+After setting the options above, change the `/initial_tf` (in the `visualization options`) to `/current_tf` to measure the difference after the calibration.
+
+The images below show that with the calibrated transformation, the projected pointcloud aligns correctly with the image.
+
+<table>
+  <tr>
+    <td><img src="../images/tag_based_pnp_calibrator/init_tf.jpg" alt="init_tf" width = 500px ></td>
+    <td><img src="../images/tag_based_pnp_calibrator/calibrated_tf.jpg" alt="calibrated_tf" width = 500px ></td>
+   </tr>
+   <tr>
+    <td><p style="text-align: center;">Before Calibration.</p></td>
+    <td><p style="text-align: center;">After Calibration.</p></td>
+  </tr>
+</table>
+
+## FAQ
+
+- Why doesn't the calibrator add calibration pairs?
```

Comment:
I missed this one before, but I think you know what is wrong here (this error is repeated a few times afterward but I will not make additional comments)

Reply:
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/2e2c0fb28cb679f2ddc21067dc30a73a8692dcb4

Reply:
You did not fix the real error...

Reply:
from your comment, I thought you were saying that the calibration, calibrator... etc shouldn't be that close to each other that is why I modify in this way.
Could you tell me what do you think it is wrong here

Reply:
`doesn't`

Reply:
Thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/5c6bbfcc0299eafff22a07a3a9e9dcf0b2914f31